dask/dataframe/shuffle.py

```diff
@@ -300,3 +243,244 @@ def collect(group, p, meta, barrier_token):
+        k = int(math.ceil(n ** (1 / stages)))
+    else:
+        k = n
+
+    groups = []
+    splits = []
+    joins = []
+
+    inputs = [tuple(digit(i, j, k) for j in range(stages))
+              for i in range(n)]
+
+    sinputs = set(inputs)
+
+    if np.isscalar(index):
+        meta = shuffle_pre_partition_scalar(df._pd, index, divisions, drop)
+        meta = meta.drop(index, axis=1)
+        df2 = map_partitions(shuffle_pre_partition_scalar, meta, df, index,
+                             divisions, drop)
+    else:
+        meta = df._pd.copy()
+        meta.index = index._pd
+        df2 = map_partitions(shuffle_pre_partition_series, meta, df, index,
+                             divisions, drop)
+
+    token = token or tokenize(df, index, divisions, max_branch)
+
+    start = dict(((name + '-join-' + token, 0, inp), (df2._name, i))
+                 for i, inp in enumerate(inputs))
+
+    for stage in range(1, stages + 1):
+        group = dict(((name + '-group-' + token, stage, inp),
+                      (shuffle_group,
+                        (name + '-join-' + token, stage - 1, inp),
+                        stage - 1, k))
+                     for inp in inputs)
+
+        split = dict(((name + '-split-' + token, stage, i, inp),
+                      (dict.get, (name + '-group-' + token, stage, inp), i, {}))
+                     for i in range(k)
+                     for inp in inputs)
+
+        join = dict(((name + '-join-' + token, stage, inp),
+                     (_concat,
+                        [(name + '-split-' + token, stage, inp[stage-1],
+                          insert(inp, stage - 1, j)) for j in range(k)
+                          if insert(inp, stage - 1, j) in sinputs]))
+                     for inp in inputs)
+        groups.append(group)
+        splits.append(split)
+        joins.append(join)
+
+    if np.isscalar(index):
+        end = dict(((name + '-' + token, i),
+                    (shuffle_post_scalar, (name + '-join-' + token, stages, inp),
+                                          index))
+                    for i, inp in enumerate(inputs))
+    else:
+        end = dict(((name + '-' + token, i),
+                    (shuffle_post_series, (name + '-join-' + token, stages, inp),
+                                          index.name))
+                    for i, inp in enumerate(inputs))
+
+    dsk = merge(df2.dask, start, end, *(groups + splits + joins))
+
+    meta = df._pd.set_index(index if np.isscalar(index) else index._pd)
+    return DataFrame(dsk, name + '-' + token, meta, divisions)
+
+
+def digit(n, k, base):
+    """
+
+    >>> digit(1234, 0, 10)
+    4
+    >>> digit(1234, 1, 10)
+    3
+    >>> digit(1234, 2, 10)
+    2
+    >>> digit(1234, 3, 10)
+    1
+    """
+    return n // base**k  % base
+
+
+def insert(tup, loc, val):
+    """
+
+    >>> insert(('a', 'b', 'c'), 0, 'x')
+    ('x', 'b', 'c')
+    """
+    L = list(tup)
+    L[loc] = val
+    return tuple(L)
```

Comment:
`digit` and `insert` are duplicated in both bag and dataframe. Perhaps move to utils?